src/gui/editorwidgets/core/qgseditorwidgetautoconf.h

```diff
@@ -0,0 +1,83 @@
+/***************************************************************************
+    qgseditorwidgetautoconf.h
+    ---------------------
+    begin                : July 2016
+    copyright            : (C) 2016 by Patrick Valsecchi
+    email                : patrick.valsecchi at camptocamp.com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSEDITORWIDGETAUTOCONF_H
+#define QGSEDITORWIDGETAUTOCONF_H
+
+#include <QList>
+#include <QSharedPointer>
+
+class QgsVectorLayer;
+class QgsWidgetSetup;
+
+/** \ingroup gui
+ * Base class for plugins allowing to pick automatically a widget type for editing fields.
+ *
+ * @note added in QGIS 3.0
+ */
+class GUI_EXPORT QgsEditorWidgetAutoConfPlugin
+{
+  public:
+    /**
+     * Typical returned values are:
+     *   * 0: no matching type found.
+     *   * 10: a widget has been guessed from the type of field.
+     *   * 20: a widget has been determined from an external configuration (for example a database table)
```

Comment:
Not if we want plugins to be able to put values in between. For example a plugin that would guess the widget to use in function of the field name would return 15 to be taken in priority compared to the one guessing from the field type but not override what is taken from external configuration.
An enum wouldn't give us as much flexibility.